proposals/extensions/android-parcelable.md

```diff
@@ -0,0 +1,262 @@
+  - What about making `Parceler` a `private object`?
+- We may want to require the object to be annotated to explicitly show that the `newArray()` is auto-generated
+
+> Discussion: why not implement `newArray()` in the `Parceler` interface itself?
+First, this method can not be implemented generically (with erasure), because the runtime type of the array created is different every time.
+We could do something like `fun newArray(size: Int) = throw UnsupportedOperationException("This method must be overridden by subclasses")`, this has the benefit of making the IDE's life easier: otherwise we'd have to teach it to skip `newArray()` when generating stubs for the Override/Implement action in annotated classes. OTOH, this is error-prone in the case of non-annotated classes. A solution here could be to magically implement `newArray()` in all concrete subclasses of `Parceler` regardless of the annotation. I wonder if this can be promoted to a general feature of Kotlin...       
+
+## Per-property and per-type Parcelers
+
+Existing annotation processors allow for per-field and per-type customization of serialization logic, e.g.:
+
+```kotlin
+@Parcelize
+class MyParcelable(
+    @CustomParceler(FooParceler::class)
+    val foo: Foo,
+    @CustomParceler(Bar.Parceler::class)
+    val bar: Bar
+): Parcelable
+``` 
+
+The `@CustomParceler` annotation can be defined as follows:
+
+```kotlin
+annotation class CustomParceler(val parcelerClass: KClass<out Parceler<*>>)
+```
+
+Additional rules:
+- the class passed as a custom Parceler for a property of type `T` must implement `Parceler<T>` and be a singleton (declared as object)
+
+Question: how to handle `describeContents()` here? Should we bitwise-or all the parcelers? How do annotation processors go about it?    
+
+It would also be desirable to provide bulk customization for all properties of the same type, e.g. `java.util.Date`. This can be done globally, e.g. though a meta-annotation, or locally, e.g.:
+
+```kotlin
+@Parcelize
+@CustomParcelerForType(
+    type = Date::class,
+    parceler = DateParceler::class
+)
+class MyParcelable(val from: Date, val to: Date): Parcelable
+```
+
+Local customization is more flexible, but will likely result in a lot of duplication.
+Global customization is problematic wrt inceremental and separate compilation, but this can be addressed in the future.  
+
+## Handling hierarchies of Parcelable classes
+
+Hierarchies of Parcelable classes are challenging from two points of view:
+- organizing constructors (see above)
+- deserializing an instance of the right class from a Parcel, when the container has only a reference to a superclass.
+
+The latter issue seems to be resolvable by "tagging" such records through writing fully-qualified names of concrete classes to parcels, but it's not clear whether it's a good idea, since
+- it involves reflection
+- it will increase the size of parcels
+
+This proposal does not support hierarchies for now.  
```

Comment:
@Takhion Well, the problem is that the primary constructor of the `@Parcelize` class should contain all significant data (all parameters of the primary constructor should be `val` or `var`). If your class hierarchy satisfies this requirement, it can be supported.